Dockerfile

```diff
@@ -1,22 +1,29 @@
-FROM pindar/jq as jq
+# === Build builder image ===
 
-FROM gradle:6.8.3-jdk11 
+FROM gradle:6.8.3-jdk11 AS build
 
-COPY --from=jq /usr/local/bin/jq /usr/local/bin/jq
-RUN chmod +x /usr/local/bin/jq
+WORKDIR /home/builder
 
-COPY lib/build.gradle /opt/test-runner/lib/build.gradle
+# Prepare required project files
+COPY lib/src ./src
+COPY build.gradle ./
 
-COPY lib/buildSrc/build.gradle /opt/test-runner/lib/buildSrc/build.gradle
-COPY lib/buildSrc/src /opt/test-runner/lib/buildSrc/src
+# Build test runner and copy cached dependencies
+RUN gradle --no-daemon -i shadowJar \
+    && cp build/libs/autotest-runner.jar . \
+    && mkdir -p /opt/test-runner/gradle \
+    && cp -r .gradle/ /opt/test-runner/gradle/.gradle/ \
+    && unlink /root/.gradle \
+    && ln -s /opt/test-runner/gradle/.gradle /root/.gradle
```

Comment:
I don't think this caching bit is working at you intend it to. The gradle files are cached within the `build` image, but they are not copied to the resulting image used below. For that, you'd need to do something like:

```Dockerfile
COPY --from=build /home/builder/.gradle .
```

(you also need to move some of the other lines setting up the caching to the runtime image)

See the [C# test runner for an example](https://github.com/exercism/csharp-test-runner/blob/main/Dockerfile#L21).

Reply:
I *think* I fixed this? This part is a bit of the black magic here that was pre-existing and I am not entirely sure that it works.

Reply:
Apparently this is not working, I am still iterating on attempts to get this resolved:

```
Execution failed for task ':compileTestJava'.
> Could not resolve all files for configuration ':testCompileClasspath'.
   > Could not resolve junit:junit:4.13.
     Required by:
         project :
      > No cached version of junit:junit:4.13 available for offline mode.
   > Could not resolve org.assertj:assertj-core:3.15.0.
     Required by:
         project :
      > No cached version of org.assertj:assertj-core:3.15.0 available for offline mode.
```